src/EventListener/InitializeSystemListener.php

```diff
@@ -275,29 +274,32 @@ class_alias('Contao\\ModuleLoader', 'ModuleLoader');
      */
     private function setRelativePath($basePath)
     {
-        Environment::set('path', $basePath);
-
-        define('TL_PATH', Environment::get('path')); // backwards compatibility
+        define('TL_PATH', $basePath); // backwards compatibility
     }
 
     /**
      * Sets the default language.
+     *
+     * @param Request $request
      */
-    private function setDefaultLanguage()
+    private function setDefaultLanguage(Request $request = null)
     {
-        if (!isset($_SESSION['TL_LANGUAGE'])) {
-            $langs = Environment::get('httpAcceptLanguage');
+        if (!$this->session->has('TL_LANGUAGE')) {
+            $langs = null !== $request ? $request->getLanguages() : [];
             array_push($langs, 'en'); // see #6533
 
             foreach ($langs as $lang) {
                 if (is_dir(__DIR__ . '/../../src/Resources/contao/languages/' . str_replace('-', '_', $lang))) {
-                    $_SESSION['TL_LANGUAGE'] = $lang;
+                    $_SESSION['TL_LANGUAGE'] = $lang; // backwards compatibility
+                    $this->session->set('TL_LANGUAGE', $lang);
                     break;
                 }
             }
         }
 
-        $GLOBALS['TL_LANGUAGE'] = $_SESSION['TL_LANGUAGE'];
+        if ($this->session->has('TL_LANGUAGE')) {
```

Comment:
The if block is unnecessary, because `$langs` always includes `en` and `src/Resources/contao/languages/en` always exists. Therefore `$this->session->has('TL_LANGUAGE')` should always be true.